README.md
# `scrape-twitter` [![Build Status](https://travis-ci.org/sebinsua/scrape-twitter.png)](https://travis-ci.org/sebinsua/scrape-twitter) [![npm version](https://badge.fury.io/js/scrape-twitter.svg)](https://npmjs.org/package/scrape-twitter)
> 🐦 Access Twitter data without an API key

This module provides command line interfaces to scrape: profiles, timelines, connections, likes, search and conversations.

It also exposes both streams and a promise returning function to help accessing Twitter in your own applications.

Real-time firehoses can be created using the companion module [`monitor-head-stream`](https://github.com/sebinsua/monitor-head-stream#example). 

## Features

- [x] Get Twitter data without being required to configure an API key.
- [x] Twitter can't constrain access as easily as they can to an API or an individual API key. Any constraints introduced would apply to their public site. A scraper can be fixed; you are no longer beholden to Twitter.
- [x] Grab timelines, whole conversations, profiles, connections, likes, etc.
- [x] Automatically pages to fetch all tweets.
- [x] Provides metadata on how tweet replies are linked together. *e.g. `isReplyToId`*

## Example

### Get profile

```sh
$ scrape-twitter profile sebinsua
# ...
```

### Get timeline

```sh
$ scrape-twitter timeline nouswaves
# ...
```

### Get likes

This command requires a valid login. It will check for the following environment variables: `TWITTER_USERNAME`, `TWITTER_PASSWORD`, `TWITTER_KDT`. But can also pick these up from a [`dotenv`](https://github.com/motdotla/dotenv) file at the path `~/.scrape-twitter`. The first time you login you will be asked to store the `TWITTER_KDT` - this is used by Twitter to recognise your device.

```sh
$ scrape-twitter likes sebinsua
# ...
```

### Get connections

This command also requires a valid login.

```sh
$ scrape-twitter connections sebinsua --type=following
# ...
```

### Get conversation

```sh
$ scrape-twitter conversation ctbeiser 691766715835924484
# ...
```

### Search

```sh
$ scrape-twitter search --query "from:afoolswisdom motivation" --type latest
# ...
```

### Get list

```sh
$ scrape-twitter list nouswaves list
# ...
```

#### JSON interface plays nicely with CLI tools like [`jq`](https://github.com/stedolan/jq), [`coreutils/gshuf`](https://github.com/wertarbyte/coreutils) and [`terminal-notifier`](https://github.com/julienXX/terminal-notifier)

For example, a [MOTD-like](https://en.wikipedia.org/wiki/Motd_(Unix)) script might contain:

```sh
scrape-twitter search --query="from:afoolswisdom knowledge" | jq -r '.[].text' | gshuf -n 1 | terminal-notifier -title "Knowledge (MOTD)"
```

## Install

*With `yarn`*:
```sh
yarn global add scrape-twitter
```

*With `npm`*:
```sh
npm install -g scrape-twitter
```

## API

### `new TimelineStream(username: string, { retweets: boolean, replies: boolean, count: ?number })`

Create a `ReadableStream<Tweet>` for the timeline of a `username`.

### `new LikeStream(username: string, { count: ?number, env: process.env })`

Create a `ReadableStream<Tweet>` for the likes of a `username`.

### `new ConnectionStream(username: string, type: 'following' | 'followers', process.env)`

Create a `ReadableStream<UserConnection>` for the connections of a `username`.

### `new ConversationStream(username: string, id: string, { count: ?number })`

Create a `ReadableStream<Tweet>` for the conversation that belongs to a `username` and tweet `id`.

### `new ThreadedConversationStream(id: string)`

Create a `ReadableStream<Tweet>` for the thread that belongs to a tweet `id`.

### `new TweetStream(query: string, type: 'top' | 'latest', { count: ?number })`

Create a `ReadableStream<Tweet>` for the tweets that match a `query` and `type`.

### `new ListStream(username: string, list: string, { count: ?number })`

Create a `ReadableStream<Tweet>` for the `username`'s `list`.

### `getUserProfile(username: string)`

Get a `Promise<UserProfile>` for a particular `username`.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.11.1"></a>
## [0.11.1](https://github.com/sebinsua/scrape-twitter/compare/v0.11.0...v0.11.1) (2018-08-29)


### Bug Fixes

* **login:** url-encode password to send correct request's body ([#19](https://github.com/sebinsua/scrape-twitter/issues/19)) ([c7535ee](https://github.com/sebinsua/scrape-twitter/commit/c7535ee))
* update parser to use data attribute for counts ([#16](https://github.com/sebinsua/scrape-twitter/issues/16)) ([00dda31](https://github.com/sebinsua/scrape-twitter/commit/00dda31))



<a name="0.11.0"></a>
# [0.11.0](https://github.com/sebinsua/scrape-twitter/compare/v0.10.0...v0.11.0) (2018-02-03)


### Features

* **timeout:** added SCRAPE_TWITTER_TIMEOUT environment variable ([b6d0f9e](https://github.com/sebinsua/scrape-twitter/commit/b6d0f9e))



<a name="0.10.0"></a>
# [0.10.0](https://github.com/sebinsua/scrape-twitter/compare/v0.8.8...v0.10.0) (2018-02-03)


### Bug Fixes

* **list:** fixed list twitter query ([817f5d5](https://github.com/sebinsua/scrape-twitter/commit/817f5d5))
* **test:** an extra tweet appears against a conversation; erm twitter? ([c31e905](https://github.com/sebinsua/scrape-twitter/commit/c31e905))


### Features

* **fetch:** https.globalAgent added to fetcher to support proxies ([a27abe2](https://github.com/sebinsua/scrape-twitter/commit/a27abe2))
* **profile:** add backgroundImage to profile ([61ca4a9](https://github.com/sebinsua/scrape-twitter/commit/61ca4a9))
* **timeline:** added media timeline ([3562531](https://github.com/sebinsua/scrape-twitter/commit/3562531))



<a name="0.9.0"></a>
# [0.9.0](https://github.com/sebinsua/scrape-twitter/compare/v0.8.8...v0.9.0) (2017-06-29)


### Bug Fixes

* **list:** fixed list twitter query ([817f5d5](https://github.com/sebinsua/scrape-twitter/commit/817f5d5))


### Features

* **fetch:** https.globalAgent added to fetcher to support proxies ([a27abe2](https://github.com/sebinsua/scrape-twitter/commit/a27abe2))
* **profile:** add backgroundImage to profile ([61ca4a9](https://github.com/sebinsua/scrape-twitter/commit/61ca4a9))



<a name="0.8.9"></a>
## [0.8.9](https://github.com/sebinsua/scrape-twitter/compare/v0.8.8...v0.8.9) (2017-05-02)


### Bug Fixes

* **test:** an extra tweet appears against a conversation; erm twitter? ([c31e905](https://github.com/sebinsua/scrape-twitter/commit/c31e905))



<a name="0.8.8"></a>
## [0.8.8](https://github.com/sebinsua/scrape-twitter/compare/v0.8.7...v0.8.8) (2017-03-30)


### Bug Fixes

* **test:** somebody deleted all of their tweets ([579690c](https://github.com/sebinsua/scrape-twitter/commit/579690c))
* **user-profile:** screenName should be picked up correctly ([8641735](https://github.com/sebinsua/scrape-twitter/commit/8641735))



<a name="0.8.7"></a>
## [0.8.7](https://github.com/sebinsua/scrape-twitter/compare/v0.8.6...v0.8.7) (2017-02-22)


### Bug Fixes

* **timeline:** last tweet should not include any retweets ([63e1e93](https://github.com/sebinsua/scrape-twitter/commit/63e1e93))



<a name="0.8.6"></a>
## [0.8.6](https://github.com/sebinsua/scrape-twitter/compare/v0.8.5...v0.8.6) (2017-02-11)



<a name="0.8.5"></a>
## [0.8.5](https://github.com/sebinsua/scrape-twitter/compare/v0.8.4...v0.8.5) (2017-02-11)


### Bug Fixes

* **check-status:** detect missing login ([7a26426](https://github.com/sebinsua/scrape-twitter/commit/7a26426))
* **test:** tweet can contain quote ([f58a9e0](https://github.com/sebinsua/scrape-twitter/commit/f58a9e0))
* **timeline:** API for replies now requires login; needs refactor ([d7cad67](https://github.com/sebinsua/scrape-twitter/commit/d7cad67))



<a name="0.8.4"></a>
## [0.8.4](https://github.com/sebinsua/scrape-twitter/compare/v0.8.3...v0.8.4) (2017-02-02)


### Bug Fixes

* **username:** treat _ as string [#2](https://github.com/sebinsua/scrape-twitter/issues/2) ([1652c26](https://github.com/sebinsua/scrape-twitter/commit/1652c26))



<a name="0.8.3"></a>
## [0.8.3](https://github.com/sebinsua/scrape-twitter/compare/v0.8.2...v0.8.3) (2017-02-02)


### Bug Fixes

* **username:** treat username as string [#2](https://github.com/sebinsua/scrape-twitter/issues/2) ([0e12174](https://github.com/sebinsua/scrape-twitter/commit/0e12174))



<a name="0.8.2"></a>
## [0.8.2](https://github.com/sebinsua/scrape-twitter/compare/v0.8.1...v0.8.2) (2017-01-27)


### Bug Fixes

* **package:** expose dist ([b62216f](https://github.com/sebinsua/scrape-twitter/commit/b62216f))



<a name="0.8.1"></a>
## [0.8.1](https://github.com/sebinsua/scrape-twitter/compare/v0.8.0...v0.8.1) (2017-01-26)


### Bug Fixes

* **connections:** connections cli is accessible ([292942e](https://github.com/sebinsua/scrape-twitter/commit/292942e))



<a name="0.8.0"></a>
# [0.8.0](https://github.com/sebinsua/scrape-twitter/compare/v0.7.2...v0.8.0) (2017-01-26)


### Features

* **connections:** added connections stream ([a1f54f5](https://github.com/sebinsua/scrape-twitter/commit/a1f54f5))



<a name="0.7.2"></a>
## [0.7.2](https://github.com/sebinsua/scrape-twitter/compare/v0.7.1...v0.7.2) (2017-01-05)


### Bug Fixes

* **env:** touch env file automatically ([626639b](https://github.com/sebinsua/scrape-twitter/commit/626639b))



<a name="0.7.1"></a>
## [0.7.1](https://github.com/sebinsua/scrape-twitter/compare/v0.7.0...v0.7.1) (2017-01-05)


### Bug Fixes

* **likes:** exposed cli ([73ab161](https://github.com/sebinsua/scrape-twitter/commit/73ab161))



<a name="0.7.0"></a>
# [0.7.0](https://github.com/sebinsua/scrape-twitter/compare/v0.6.8...v0.7.0) (2017-01-05)


### Bug Fixes

* **likes:** better description of usage ([6be8030](https://github.com/sebinsua/scrape-twitter/commit/6be8030))


### Features

* **likes:** get twitter likes ([5583bf4](https://github.com/sebinsua/scrape-twitter/commit/5583bf4))



<a name="0.6.8"></a>

cook.js
#!/usr/bin/env node

var cheerio = require('cheerio')

// Place default kdt into tough-cookie.
var expandHome = require('expand-home-dir')
require('dotenv').config({ path: expandHome('~/.scrape-twitter') })

var twitterUsername = process.env.TWITTER_USERNAME
var twitterPassword = process.env.TWITTER_PASSWORD
if (!twitterUsername || !twitterPassword) {
  process.exit(1)
}

var fetchDecorator = require('fetch-cookie')
var isomorphicFetch = require('isomorphic-fetch')
var tough = require('tough-cookie')
var denodeify = require('es6-denodeify')(Promise)

function getFetchWithKdt () {
  var jar = new tough.CookieJar()
  var setCookie = denodeify(jar.setCookie.bind(jar))

  var fetch = fetchDecorator(isomorphicFetch, jar)
  if (process.env.TWITTER_KDT) {
    var cookie = 'kdt=' + process.env.TWITTER_KDT + '; Expires=Fri, 06 Jul 2018 02:30:12 GMT; Path=/; Domain=.twitter.com; Secure; HTTPOnly'
    var url = 'https://twitter.com/sessions'
    return setCookie(cookie, url).then(function () {
      return fetch
    })
  } else {
    return Promise.resolve().then(function () {
      return fetch
    })
  }
}

var toText = function (response) { return response.text() }

getFetchWithKdt().then(function (fetch) {
  fetch('https://twitter.com').then(toText).then(function (body) {
    var $ = cheerio.load(body)
    var authToken = $('input[name="authenticity_token"]').val()

    var formData = 'session%5Busername_or_email%5D=' + twitterUsername + ' &session%5Bpassword%5D=' + twitterPassword + '&remember_me=1&return_to_ssl=true&scribe_log=&redirect_after_login=%2F&authenticity_token=' + authToken
    return fetch('https://twitter.com/sessions', {
      method: 'POST',
      redirect: 'manual',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:50.0) Gecko/20100101 Firefox/50.0'
      },
      body: formData
    }).then(function (res) {
      // TODO: Store the KDT.
      // var getCookieString = denodeify(jar.getCookieString.bind(jar))
      return fetch('https://twitter.com/i/likes')
        .then(toText)
        .then(function (_body) {
          var $ = cheerio.load(_body)
          console.log($('.original-tweet:first-child .tweet-text').text())
        })
    })
  })
})
